File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/shiftcoverage/ShiftCoverage.kt
package com.reguerta.user.domain.shiftcoverage

import kotlinx.serialization.Serializable

@Serializable
internal data class ShiftCoverageSnapshot(
    val schemaVersion: Int,
    val environment: String,
    val memberId: String,
    val isAdmin: Boolean,
    val eligible: Boolean,
    val serverTimeMillis: Long,
    val policyRevision: String,
    val policy: Policy,
    val cases: List<Case>,
    val credits: List<Credit>,
    val reserves: List<Reserve>,
    val availableShifts: List<AvailableShift> = emptyList(),
    val members: List<MemberLabel> = emptyList(),
    val notifications: List<Notification> = emptyList(),
    val notification: NotificationReference? = null,
) {
    @Serializable
    data class Notification(val eventId: String, val sentAtMillis: Long)

    @Serializable
    data class NotificationReference(val eventId: String, val caseId: String, val caseRevision: Long)

    @Serializable
    enum class Kind { delivery, market }
    @Serializable
    enum class Status { open, offered, accepted, completed, cancelled, failed }
    @Serializable
    enum class Phase { reserve, volunteers, drawRequired, draw, adminRequired }
    @Serializable
    enum class OfferSource { admin, reserve, volunteer, draw }
    @Serializable
    enum class CreditState { pending, consumed }

    @Serializable
    data class Policy(
        val maximumOfferWindowMillis: Long,
        val volunteerWindowMillis: Long?,
        val drawAvailable: Boolean,
    )

    @Serializable
    data class Case(
        val caseId: String,
        val shiftId: String,
        val type: Kind,
        val positionIndex: Int,
        val status: Status,
        val revision: Long,
        val shiftRevision: Long,
        val scheduledAtMillis: Long,
        val writable: Boolean,
        val absentUserId: String,
        val acceptedUserId: String?,
        val offer: Offer?,
        val selectionPhase: Phase?,
        val volunteerClosesAtMillis: Long?,
        val volunteered: Boolean,
        val updatedAtMillis: Long,
        val administration: Administration?,
        val openedByMe: Boolean = false,
        val canResumeAdmin: Boolean = false,
        val hasVolunteered: Boolean = false,
        val drawCommitted: Boolean = false,
        val drawAvailableAtMillis: Long? = null,
    )

    @Serializable
    data class Offer(val userId: String, val source: OfferSource, val expiresAtMillis: Long)

    @Serializable
    data class Administration(val openedByUserId: String, val reason: String, val volunteerCount: Int)

    @Serializable
    data class Credit(
        val creditId: String,
        val shiftId: String,
        val type: Kind,
        val state: CreditState,
        val earnedAtMillis: Long,
        val consumedAtMillis: Long?,
    )

    @Serializable
    data class MemberLabel(val memberId: String, val displayName: String, val offerCandidate: Boolean = false)

    @Serializable
    data class AvailableShift(
        val shiftId: String,
        val type: Kind,
        val scheduledAtMillis: Long,
        val shiftRevision: Long,
        val writable: Boolean,
        val assignedUserIds: List<String>,
    )

    @Serializable
    data class Reserve(val type: Kind, val active: Boolean, val enteredAtMillis: Long)
}

/** Change the revision on logout/relogin, UID, member, environment or authorization changes. */
internal data class ShiftCoverageSession(val uid: String, val memberId: String, val authorizationRevision: Long)

/** Preserve this complete value for explicit replay after an uncertain response. Never put actor identity here. */
@Serializable
internal data class ShiftCoverageCommand(
    val caseId: String,
    val operationId: String,
    val expectedRevision: Long,
    val expectedShiftRevision: Long,
    val action: Action,
    val shiftId: String? = null,
    val absentUserId: String? = null,
    val reason: String? = null,
    val userId: String? = null,
    val expiresAtMillis: Long? = null,
) {
    val schemaVersion: Int = 1
    val environment: String = "develop"

    @Serializable
    enum class Action {
        open, offer, accept, decline, expire, complete, startSelection, volunteer, withdrawVolunteer,
        commitDraw, revealDraw, offerNext, offerAdmin, resumeAdmin, cancel, fail,
    }
}

internal sealed class ShiftCoverageFailure : Exception() {
    data object SessionChanged : ShiftCoverageFailure()
    data object LocalOnly : ShiftCoverageFailure()
    data object InvalidResponse : ShiftCoverageFailure()
    data object Unavailable : ShiftCoverageFailure()
    data class Rejected(val status: Int, val code: String) : ShiftCoverageFailure()
}

internal interface ShiftCoverageRepository {
    suspend fun read(caseId: String?, session: ShiftCoverageSession): ShiftCoverageSnapshot
    suspend fun readNotification(eventId: String, session: ShiftCoverageSession): ShiftCoverageSnapshot =
        throw ShiftCoverageFailure.Unavailable
    suspend fun execute(command: ShiftCoverageCommand, session: ShiftCoverageSession)
}


internal interface CoverageRehearsalAccess {
    val repository: ShiftCoverageRepository
    suspend fun signIn(email: String, password: String): ShiftCoverageSession
    fun signOut()
}

File: functions/src/shift-coverage.ts
import type {CoverageSelection} from "./shift-coverage-selection.js";
import {HttpRequestError} from "./backend-security.js";
import {isEligibleForShiftRotation} from "./shift-eligibility.js";

export const SHIFT_COVERAGE_POLICY_REVISION = "hu084-provisional-v1" as const;

type CommandBase = {
  schemaVersion: 1;
  environment: "develop";
  caseId: string;
  operationId: string;
  expectedRevision: number;
  expectedShiftRevision: number;
};

export type ShiftCoverageCommand = CommandBase & (
  | {action: "open"; shiftId: string; absentUserId: string; reason: string}
  | {action: "offer"; userId: string; reason: string; expiresAtMillis: number}
  | {action: "accept" | "decline" | "expire" | "complete" |
      "startSelection" | "volunteer" | "withdrawVolunteer"}
  | {action: "offerNext"; expiresAtMillis: number}
  | {action: "cancel" | "fail"; reason: string}
);

export type ShiftCoverageCase = {
  schemaVersion: 1;
  policyRevision: typeof SHIFT_COVERAGE_POLICY_REVISION;
  environment: "develop";
  id: string;
  shiftId: string;
  type: "delivery" | "market";
  absentUserId: string;
  positionIndex: number;
  ownershipDigest: string;
  openedByUserId: string;
  reason: string;
  status: "open" | "offered" | "accepted" |
    "completed" | "cancelled" | "failed";
  revision: number;
  selection?: CoverageSelection;
  offer: {
    source: "admin" | "reserve" | "volunteer";
    id: string;
    userId: string;
    expiresAtMillis: number;
    assignmentContextDigest: string;
  } | null;
  acceptedUserId: string | null;
  creditId: string | null;
  createdAtMillis: number;
  updatedAtMillis: number;
};

export type ShiftCoverageCredit = {
  schemaVersion: 1;
  policyRevision: typeof SHIFT_COVERAGE_POLICY_REVISION;
  id: string;
  caseId: string;
  shiftId: string;
  type: "delivery" | "market";
  userId: string;
  state: "pending";
  earnedAtMillis: number;
  completionRevision: number;
};

export const rejectCoverage = (code: string): never => {
  throw new HttpRequestError(409, code, "Coverage transition rejected");
};

export const coverageId = (value: unknown): string => {
  if (typeof value !== "string" ||
      !/^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/.test(value)) {
    return rejectCoverage("invalid_coverage_identity");
  }
  return value;
};

const revision = (value: unknown): number => {
  if (!Number.isSafeInteger(value) || (value as number) < 0 ||
      value === Number.MAX_SAFE_INTEGER) {
    return rejectCoverage("invalid_coverage_revision");
  }
  return value as number;
};

const reason = (value: unknown): string => {
  if (typeof value !== "string" || !value.trim() || value.length > 500) {
    return rejectCoverage("invalid_coverage_reason");
  }
  return value.trim();
};

/**
 * Provisional develop command; actor identity is never accepted from this body.
 * Unknown fields reject so callers cannot smuggle completion or credit values.
 * @param {unknown} value Untrusted command body.
 * @return {ShiftCoverageCommand} Detached, exact command.
 */
export const parseShiftCoverageCommand = (
  value: unknown,
): ShiftCoverageCommand => {
  if (!value || typeof value !== "object" ||
      Object.getPrototypeOf(value) !== Object.prototype) {
    return rejectCoverage("invalid_coverage_command");
  }
  const body = value as Record<string, unknown>;
  const base = {
    schemaVersion: 1 as const,
    environment: "develop" as const,
    caseId: coverageId(body.caseId),
    operationId: coverageId(body.operationId),
    expectedRevision: revision(body.expectedRevision),
    expectedShiftRevision: revision(body.expectedShiftRevision),
  };
  let command: ShiftCoverageCommand;
  switch (body.action) {
  case "open":
    command = {...base, action: "open", shiftId: coverageId(body.shiftId),
      absentUserId: coverageId(body.absentUserId), reason: reason(body.reason)};
    break;
  case "offer":
    command = {...base, action: "offer", userId: coverageId(body.userId),
      reason: reason(body.reason),
      expiresAtMillis: revision(body.expiresAtMillis)};
    break;
  case "offerNext":
    command = {...base, action: "offerNext",
      expiresAtMillis: revision(body.expiresAtMillis)};
    break;
  case "startSelection": case "volunteer": case "withdrawVolunteer":
  case "accept": case "decline": case "expire": case "complete":
    command = {...base, action: body.action};
    break;
  case "cancel": case "fail":
    command = {...base, action: body.action, reason: reason(body.reason)};
    break;
  default:
    return rejectCoverage("invalid_coverage_action");
  }
  if (body.schemaVersion !== 1 || body.environment !== "develop" ||
      Object.keys(body).length !== Object.keys(command).length ||
      Object.keys(body).some((key) => !(key in command)) ||
      (command.action === "open" && command.expectedRevision !== 0)) {
    return rejectCoverage("invalid_coverage_command");
  }
  return command;
};

/**
 * Reads the same eligibility predicate as normal rotation. Malformed membership
 * fails closed; an administrator role never makes a real producer eligible.
 * @param {unknown} value Current transactionally read user document.
 * @return {object} Actor role and coverage eligibility.
 */
export const coverageMember = (value: unknown) => {
  const member = value as Record<string, unknown> | undefined;
  if (!member || typeof member.isActive !== "boolean" ||
      typeof member.isCommonPurchaseManager !== "boolean" ||
      !Array.isArray(member.roles) ||
      !member.roles.every((role) => typeof role === "string")) {
    return rejectCoverage("invalid_coverage_member");
  }
  const roles = member.roles.map((role: string) => role.trim().toLowerCase());
  return {
    active: member.isActive,
    admin: member.isActive && roles.includes("admin"),
    eligible: isEligibleForShiftRotation({
      isActive: member.isActive, roles,
      isCommonPurchaseManager: member.isCommonPurchaseManager,
    }),
  };
};
